enhancements/network/sdn-live-migration.md

```diff
@@ -0,0 +1,358 @@
+
+2. CNO will label each existing node with, eg:
+
+    ```yaml
+    migration.network.openshift.io/plugin: OpenShiftSDN
+    k8s.ovn.org/hybrid-overlay-node-subnet: 10.129.0.0/23
+    ```
+
+   (CNO knows how to figure out the host subnet by looking at openshift-sdn
+   HostSubnet objects. When migrating in the reverse direction, it knows to look
+   at the ovn-kubernetes annotations.)
+
+3. CNO will redeploy the openshift-sdn DaemonSets, enabling “migration mode”,
+   and adding node affinity to the node DaemonSet:
+    
+    ```yaml
+    affinity:
+      nodeAffinity:
+        requiredDuringSchedulingIgnoredDuringExecution:
+          nodeSelectorTerms:
+            - matchExpressions:
+                - key: migration.network.openshift.io/plugin
+                  operator: In
+                  values: ["OpenShiftSDN"]
+    ```
+
+  so that openshift-sdn will now only run on nodes labeled
+  “migration.network.openshift.io/plugin: OpenShiftSDN” (which is currently all
+  of them).
+
+4. CNO will then deploy the ovnkube-master and ovnkube-node DaemonSets, also
+   enabling hybrid overlay mode” in their config, and adding equivalent node
+   affinity to the ovnkube-node DaemonSet (to only run on nodes labeled
+   “migration.network.openshift.io/plugin: OVNKubernetes”). Since there are not
+   currently any nodes labeled that way, this means no actual ovnkube-node pods
+   will be created at this step.
+
+At this point, neither plugin is willing to run on a node that has no
+“migration.network.openshift.io/plugin” label at all. This means that if a new
+node gets added to the cluster in the middle of the migration process, neither
+plugin will get deployed to it until after the migration of the older nodes is
+complete, and the node affinity is removed.
+
+#### Migration
+
+In order to run ovn-kubernetess in shared-gw mode, we need MCO to update the
+ovs-configuration.service. To apply the MachineConfig update, a reboot will be
+triggered by MCO. To coordinate the draining and rebooting between CNO and MCO,
+we need to modify MCO as following:
+- add a `machineconfiguration.openshift.io/rebootPausedBy: network-operator`
+  node annotation to MCO, so that CNO can block MCO from rebooting the node
+- add a new value `RebootPending` to the existing
+  `machineconfiguration.openshift.io/state` node annotation. It will be set by
+  MCD if the reboot is blocked. It indicts that draining node is done, and the
+  rebooting is paused.
+
+When we kick off the migration from CNO, CNO will update the
+`status.networkType` field of the `network.config` CR. It will trigger MCO to
+apply new MachineConfig to each node:
+1. MCO will 
+   - rerender the MachineConfig for each node
+   - try to cordon and drain the node one node
+2. CNO finds a node is being drained by MCO by watching the MCO node
+   annotations, it will:
+   - pause the node reboot by annotate the node with
+     `machineconfiguration.openshift.io/rebootPausedBy: network-operator`.
```

Comment:
Risk here with watching for a drain vs proactively annotating nodes.  Possibility if there are issues in CNO the pause annotation won't be set before reboot.  Can the annotation be set before this kicks off?

Reply:
I refactored the process, PTAL.